### PR TITLE
Feature/speedup random sampling filter

### PR DESCRIPTION
**Pull request by YoshuaNava**
In order to resolve issue #393 I worked on speeding up the RandomSamplingDatapointsFilter. For that I implemented two random sampling methods: 

1) Direct RNG, which works similar to the previous implementation, directly querying a random number engine (int), which is transformed to a float in the range 0-1. Result: ~1.5-2x speedup
2) Uniform, which queries a uniform distribution between 0 and 1. Result: ~30% speedup.

I chose to sample directly a set of floats instead of templating that type. The reason being that float arithmetic should be enough, in my view, to achieve the goal of the filter. Double precision numbers might introduce an overhead.

I refactored a bit the existing unit testing code and ran it locally.

Last but not least, I tried to use the new std::random library, as recommended [here](https://codeforces.com/blog/entry/61587). Related [talk](https://channel9.msdn.com/Events/GoingNative/2013/rand-Considered-Harmful).

I'm submitting this PR for early feedback. Nonetheless, I still plan to run and plot the samples to get a grasp on the quality of the sampling process, as requested by @pomerlef in the related issue :slightly_smiling_face:

### CODE REVIEW

**Comment by pomerlef** [comment]
add to whitelist

**Comment by YoshuaNava** [comment]
The build is faling. Do you know what might be the reason?

**Comment by pomerlef** [comment]
It's the macos built that failed:

```bash
/Users/macmini-admin/workspace/pointmatcher/label/osx/pointmatcher/DataPointsFilters/RandomSampling.cpp:91:14: error: no type named 'Index' in namespace 'Eigen'
        for (Eigen::Index i = 0; i < nbPointsIn && j<=nbPointsOut; i++)
             ~~~~~~~^
1 error generated.

make[2]: *** [CMakeFiles/pointmatcher.dir/pointmatcher/DataPointsFilters/RandomSampling.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
2 warnings generated.

make[1]: *** [CMakeFiles/pointmatcher.dir/all] Error 2

make: *** [all] Error 2
```

**Comment by YoshuaNava** [comment]
Sorry for the wait, I had switched to other tasks at work.

I modified the file `filterProfiler.cpp` in the examples folder so that I could run some tests of the random sampling filter. I also added code to the filter to export the random indices matrix to a file.

This is what we get when we use the 'Direct RNG' sampling method:

direct_rng


And this is what we get when we enforce uniform sampling:

uniform


I used the file car_cloud400.csv for the tests.

**Comment by pomerlef** [comment]
Looking good!

**Comment by pomerlef** [comment]
most probably the same compilation error

File: pointmatcher/DataPointsFilters/RandomSampling.cpp [hunk 1]
```diff
@@ -53,24 +56,46 @@ RandomSamplingDataPointsFilter<T>::filter(const DataPoints& input)
...
 void RandomSamplingDataPointsFilter<T>::inPlaceFilter(
 	DataPoints& cloud)
 {
 	const int nbPointsIn = cloud.features.cols();
+	const int nbPointsOut = nbPointsIn * prob;
 
-	int j = 0;
-	for (int i = 0; i < nbPointsIn; ++i)
+	const Eigen::VectorXf randomNumbers = sampleRandomIndices(nbPointsIn);
+	int j=0;
```
Comment (YoshuaNava) [inline pointmatcher/DataPointsFilters/RandomSampling.cpp:90]:
```suggestion
	size_t j=0;
```

File: pointmatcher/DataPointsFilters/RandomSampling.cpp [hunk 2]
```diff
@@ -53,24 +56,46 @@ RandomSamplingDataPointsFilter<T>::filter(const DataPoints& input)
...
+	for (Eigen::Index i = 0; i < nbPointsIn && j<=nbPointsOut; i++)
```
Comment (YoshuaNava) [inline pointmatcher/DataPointsFilters/RandomSampling.cpp:91]:
```suggestion
	for (size_t i = 0; i < nbPointsIn && j<=nbPointsOut; i++)
```

File: pointmatcher/DataPointsFilters/RandomSampling.cpp [hunk 3]
```diff
@@ -53,24 +56,46 @@ RandomSamplingDataPointsFilter<T>::filter(const DataPoints& input)
...
+	}
+}
+
 // In-place filter
 template<typename T>
 void RandomSamplingDataPointsFilter<T>::inPlaceFilter(
 	DataPoints& cloud)
 {
 	const int nbPointsIn = cloud.features.cols();
+	const int nbPointsOut = nbPointsIn * prob;
```
Comment (YoshuaNava) [inline pointmatcher/DataPointsFilters/RandomSampling.cpp:87]:
```suggestion
	const size_t nbPointsOut = nbPointsIn * prob;
```

File: pointmatcher/DataPointsFilters/RandomSampling.cpp [hunk 4]
```diff
@@ -53,24 +56,46 @@ RandomSamplingDataPointsFilter<T>::filter(const DataPoints& input)
...
+		}
+	}
+}
+
 // In-place filter
 template<typename T>
 void RandomSamplingDataPointsFilter<T>::inPlaceFilter(
 	DataPoints& cloud)
 {
 	const int nbPointsIn = cloud.features.cols();
```
Comment (YoshuaNava) [inline pointmatcher/DataPointsFilters/RandomSampling.cpp:86]:
```suggestion
	const size_t nbPointsIn = cloud.features.cols();
```

File: pointmatcher/DataPointsFilters/RandomSampling.cpp [hunk 5]
```diff
@@ -53,24 +56,46 @@ RandomSamplingDataPointsFilter<T>::filter(const DataPoints& input)
...
+	{
+		default:	// Direct RNG.
+		{
+			const float randomNumberRange = gen.max() - gen.min();
+			return Eigen::VectorXf::NullaryExpr(nbPoints, [&](float dummy){return static_cast<float>(gen() / randomNumberRange);});
+		}
+		case 1:		// Uniform distribution.
+		{
+			std::uniform_real_distribution<float> dis(0, 1);
+			return Eigen::VectorXf::NullaryExpr(nbPoints, [&](float dummy){return dis(gen);});
```
Comment (YoshuaNava) [inline pointmatcher/DataPointsFilters/RandomSampling.cpp:76]:
```suggestion
			return Eigen::VectorXf::NullaryExpr(nbPoints, [&](float){return dis(gen);});
```

File: pointmatcher/DataPointsFilters/RandomSampling.cpp [hunk 6]
```diff
@@ -53,24 +56,46 @@ RandomSamplingDataPointsFilter<T>::filter(const DataPoints& input)
...
+{
+	std::random_device rd;
+	std::minstd_rand gen(rd());
+
+	switch(randomSamplingMethod)
+	{
+		default:	// Direct RNG.
+		{
+			const float randomNumberRange = gen.max() - gen.min();
+			return Eigen::VectorXf::NullaryExpr(nbPoints, [&](float dummy){return static_cast<float>(gen() / randomNumberRange);});
```
Comment (YoshuaNava) [inline pointmatcher/DataPointsFilters/RandomSampling.cpp:71]:
```suggestion
			return Eigen::VectorXf::NullaryExpr(nbPoints, [&](float){return static_cast<float>(gen() / randomNumberRange);});
```

**Comment by YoshuaNava** [comment]
I'll rebase berfore we can merge, to keep the history short.

**Pull request merged**